Sprint 8 solution time and duration

Ilux-i opened the pull request:
Добавлена новая логика приложения и соответствующие тесты

Ilux-i committed:
feature

Добавил 2 поля duration и startTime, обновил создание объекта Task и его преобразования в строку и из строки(В файл сохраняется и забирается без проблем) временно замучены файлы тестов по FileBackedTaskManager
Доработать логику выполнения работы наследуемых классов

Ilux-i committed:
feature

Интегрированы новые поля duration, startTime в Task, Epic и SubTask.
Добавлено автообновление полей(duration, startTime, endTime) эпика через метод checkingTheEpicExecutionTime()

Ilux-i committed:
feature

Добавлена вывод списка задач в порядке приоритета со сложностью n(0)

Ilux-i committed:
feature

реализована проверка пересечений с помощью Stream API

Ilux-i committed:
refactor

Изменены циклы типа foreach на Stream API

Ilux-i committed:
fix

Отредактирован код
Размучены тесты
Все прошлые тесты проходят проверку

Ilux-i committed:
fix and add test

Добавление тестов согласно логике работы приложения
Исправление багов
Прохождение всех тестов

Ilux-i committed:
fix

...

KirPyt commented on src/main/java/manager/FileBackedTaskManager.java:
```diff
@@ -7,13 +7,13 @@
-    private static final String HEADER = "id,type,name,status,description,epic";
+    private String file;
+    private static final String HEADER = "id,type,name,status,description,duration,startTime,moreInfo";
+    TreeSet<Task> sortedSet = new TreeSet<>(Comparator.comparing(Task::getStartTime));
```
Данное поле так же надо инкапсулировать

KirPyt replied:
А так же перенести сортированное множество в  `InMemoryTaskManager `

KirPyt commented on src/main/java/manager/FileBackedTaskManager.java:
```diff
@@ -37,8 +37,17 @@ public void clearSubTasks() {
-        super.addTask(task);
-        save();
+        try {
+            if (checkingIntersectionsForSortedSet(task)) {
```
Вы можете пробрасывать исключение непосредственно в методе `checkingIntersectionsForSortedSet`

KirPyt commented on src/main/java/manager/FileBackedTaskManager.java:
```diff
@@ -49,15 +58,40 @@ public void addEpic(Epic epic) {
     public void addSubTask(SubTask subTask) {
-        super.addSubTask(subTask);
-        save();
+        try {
```
В целом - вынесите логику проверки пересечения в `InMemoryTaskManager`, тогда все другие наследуемые менеджеры автоматически получат проверку пересечения

KirPyt commented on src/main/java/manager/InMemoryTaskManager.java:
```diff
@@ -42,27 +43,29 @@ public List<Task> getHistory() {
-        for (Task task : tasks.values()) {
-            removeTaskById(task.getId());
-        }
+        new ArrayList<>(tasks.values()).stream()
```
Оборачивать в ArrayList нет необходимости, стоит попробовать обработать возвращаемый тип `Collection`

KirPyt commented on src/main/java/task/Epic.java:
```diff
@@ -30,4 +48,29 @@ public String toString() {
+                .min(Comparator.naturalOrder());
+
+        if (endTime.isPresent()) {
+            this.duration = Duration.between(startTime.get(), endTime.get());
```
Продолжительность эпика - сумма продолжительностей подзадач

KirPyt commented on src/main/java/Main.java:
```diff
@@ -1,16 +1,18 @@
-        manager.addEpic(new Epic("epic", "description4"));
-        manager.addSubTask(new SubTask("subtask", "description4", 1));
+        FileBackedTaskManager manager = FileBackedTaskManager.loadFromFile("data_test.csv");
+//        manager.addTask(new Task("task1", "description4", Duration.ofMinutes(90L), LocalDateTime.now()));
```
Закомментированный код стоит убрать

KirPyt reviewed (commented):
Здравствуйте!

Указал замечания по коду, над которыми надо поработать. В целом из важного - перенести проверку пересечения задач в базовый менеджер `InMemoryTaskManager`, так же перенести туда работу с сортированным множеством. Перепроверьте логику расчета параметров эпика - продолжительность эпика, это сумма продолжительностей его подзадач, время окончания эпика - время окончания последней подзадачи в эпике, время начала - время начала самой ранней подзадачи

Если будут вопросы по ревью, то пишите в Пачку

Ilux-i committed:
refactor

Внесение небольших коррекировок, согласно комментариям

Ilux-i committed:
refactor

...

KirPyt commented on src/main/java/manager/FileBackedTaskManager.java:
```diff
@@ -37,8 +36,15 @@ public void clearSubTasks() {
-        super.addTask(task);
-        save();
+        try {
+            checkingIntersectionsForSortedSet(task);
```
Логику поиска пересечений надо вынести в базовый класс InMemoryTaskMamager, чтобы все наследники базового менеджера наследовали эту логику

KirPyt reviewed (commented):
Здравствуйте!

Перенесите логику поиска пересечений (вызовы в методах) в inMemoryTaskManager, в остальном все хорошо

Ilux-i committed:
refactor

Перенесена логика поиска пересечений (вызовы в методах) из FileBackedTaskManager в inMemoryTaskManager

KirPyt reviewed (commented):
Здравствуйте!

Рад, что получилось поработать над комментариями

Ваш проект зачтен, хорошего дня!
